### PR TITLE
Feat: favorites list for loggedIn users

### PR DESCRIPTION
**Pull request by kevynfg**
I had to do a little refactor for validation purposes for the "like" request popping some errors while trying to set with an undefined variable.

added new feature for fetching user favorites list.

I personally couldn't find a fitting color for the sidebar, so i went with the one that already exists on the code, if anyone have suggestions, i'm all ears.

Feel free to point corrections or code smells.

### CODE REVIEW

**Comment by matheusmisumoto** [comment]
The code itself seems good to me.

I would just point out two things for a better user experience:
- The favorites menu should appear only for logged in users.
- The like button should update the favorites list without the need to reload the page.

**Comment by kevynfg** [comment]
> The code itself seems good to me.
> 
> I would just point out two things for a better user experience:
> 
> * The favorites menu should appear only for logged in users.
> * The like button should update the favorites list without the need to reload the page.

seems good to me, i'll work on this later today, thank you.

**Comment by kevynfg** [comment]
Made a few corrections, hope everything is ok

**Comment by matheusmisumoto** [comment]
Somehow the like button adds the link twice on favorites list.
Seems like a React issue, because when the page is refreshed, the list appears normal.

**Edit:** seems that removing Strict Mode fixes the issue.

duplicate

**Comment by kevynfg** [comment]
Yes, i noticed that and removed strict mode while i was testing, perhaps on production won't happen.

**Comment by johnylab** [comment]
> Yes, i noticed that and removed strict mode while i was testing, perhaps on production won't happen.

I'm taking a look.

File: src/components/FavoritesLink/styled.js [hunk 1]
```diff
@@ -0,0 +1,20 @@
...
+    cursor: pointer;
+    transition: color 0.2s;
+
+    &:hover {
+        color: ${uiColors.hover};
+    }
+`;
+
+export const FavoritesSidebarContainer = styled.div`
+    opacity: ${({showSidebar}) => showSidebar ? 1 : 0};
```
Comment (johnylab) [inline src/components/FavoritesLink/styled.js]: This way the new panel is blocking the UI. The fav link and user menu are no clickable:

1. please move favorites link inside the user menu
2. move the favorites panel to `App` (see example below)
3. use context api to open the panel

```
function App() {
  return (
    <>
      <Header />
      <Wrapper>
        <AddLink />
        <LinkList />
      </Wrapper>
      <FavoritesPanel />
    </>
  );
}
```

image

File: src/components/FavoritesSidebar/styled.js [hunk 2]
```diff
@@ -0,0 +1,85 @@
...
+    overflow-y: auto;
+    padding: 0;
+    padding: 0;
+    margin: 0;
+    list-style: none;    
+    display: flex;
+    flex-direction: column;
+`
+
+export const FavoritesTitle = styled.h1`
```
Comment (johnylab) [inline src/components/FavoritesSidebar/styled.js]: Definitely don't use h1 here. You're defining the whole page title to this.

File: src/components/FavoritesSidebar/styled.js [hunk 3]
```diff
@@ -0,0 +1,85 @@
...
+    text-align: center;
+    color: ${grey.light};
+`
+
+export const FavoritesLinks = styled.a`
```
Comment (johnylab) [inline src/components/FavoritesSidebar/styled.js]:
```suggestion
export const FavoritesLink = styled.a`
```

File: src/components/FavoritesSidebar/index.js [hunk 4]
```diff
@@ -0,0 +1,27 @@
...
+                                href={favoriteItem.url}
+                                target="_blank"
+                                rel="noreferrer"
+                            >
+                                {favoriteItem.title || favoriteItem.url}
+                            </FavoritesLinks>
+                        </li>
+                    ))}
+                </FavoritesUl>
+        </FavoriteSidebar>
```
Comment (johnylab) [inline src/components/FavoritesSidebar/index.js]: Looks like Sidebar is made to show favorites only, while FavoritesSidebar is a reusable component, so I'd change the namings like this:

```
FavoritesPanel = (props) =>
  <Sidebar>
    <SidebarTitle />
    <SidebarContent>
      <ul>
        {likes.map((props) => <FavoriteCard {...props} />)}
      </ul>
    </SidebarContent>
  </Sidebar>
```

File: src/components/FavoritesLink/styled.js [hunk 5]
```diff
@@ -0,0 +1,20 @@
+import styled from "styled-components";
+import { uiColors } from "../../constants/colors";
+
+export const FavoritesLink = styled.ul`
```
Comment (johnylab) [inline src/components/FavoritesLink/styled.js]:
```suggestion
export const FavoritesList = styled.ul`
```

File: src/components/FavoritesSidebar/index.js [hunk 6]
```diff
@@ -0,0 +1,27 @@
+import { useContext } from "react";
+import { CloseFavoritesSidebar, FavoriteSidebar, FavoritesLinks, FavoritesTitle, FavoritesUl } from "./styled";
+import { GlobalContext } from "../../context/GlobalContext";
+
+export default function Sidebar({initialLikes, setShowSidebar}) { 
+    const { likes } = useContext(GlobalContext);
```
Comment (johnylab) [inline src/components/FavoritesSidebar/index.js]: No need to touch GlobalContext here. Firebase can update this list automaticaly.

```
import { getDatabase, ref, onChildAdded, onChildChanged, onChildRemoved } from "firebase/database";

const db = getDatabase();
const commentsRef = ref(db, 'post-comments/' + postId);
onChildAdded(commentsRef, (data) => {
  addCommentElement(postElement, data.key, data.val().text, data.val().author);
});

onChildChanged(commentsRef, (data) => {
  setCommentValues(postElement, data.key, data.val().text, data.val().author);
});

onChildRemoved(commentsRef, (data) => {
  deleteComment(postElement, data.key);
});
```

Please refer to [this documentation](https://firebase.google.com/docs/database/web/lists-of-data?hl=pt&authuser=0#web-version-9_2) for more details.

File: src/components/LinkItem/InteractionToolBar.js [hunk 7]
```diff
@@ -12,23 +12,37 @@ import { InteractionToolBarWrapper, LinkInteractionButton, LinkItemButtonLabel }
...
+                    const parsedLink = parseLike(link);
+                    unsetUserLike(parsedLink);
+                });
                 setIsLiked(false);
             } else {
-                addLinkToCollection('Likes', id, user);
+                addLinkToCollection('Likes', id, user).then((link) => {
+                    const parsedLink = parseLike(link);
+                    setUserLikes(parsedLink);
+                });
```
Comment (johnylab) [inline src/components/LinkItem/InteractionToolBar.js]: No changes are needed for this file. See comments above.

File: src/components/LinkList/index.js [hunk 8]
```diff
@@ -36,8 +38,16 @@ const LoadMore = ({ next, loading, empty, ended }) => {
...
+        next, requestSearch, setUserLikes
     } = useContext(GlobalContext);
+    const { user } = useAuth();
+    const { initialLikes } = useUserLikes(user);
+
+    useEffect(() => {
+         if (initialLikes && initialLikes.length) {
+            setUserLikes(initialLikes);
+        };
+    }, [initialLikes])
```
Comment (johnylab) [inline src/components/LinkList/index.js]: No changes are needed for this file. See comments above.

File: src/context/AppReducer.ts [hunk 9]
```diff
@@ -93,6 +102,26 @@ const AppReducer = (state: GlobalState, action: Action) => {
...
+                ...state,
+            };
+        
+        case actionTypes.UNSET_USER_LIKE:
+            const likes = state.likes.filter((like) => like.url !== action.payload.url);
+            return {
+                ...state,
+                likes,
+            }
+
```
Comment (johnylab) [inline src/context/AppReducer.ts]: No changes are needed for this file. See comments above.

File: src/context/GlobalContext.js [hunk 10]
```diff
@@ -56,6 +70,8 @@ export const GlobalProvider = ({ children }) => {
             add,
             remove,
             requestSearch,
+            setUserLikes,
+            unsetUserLike,
```
Comment (johnylab) [inline src/context/GlobalContext.js]: No changes are needed for this file. See comments above.

File: src/firebase/userCollections.js [hunk 11]
```diff
@@ -77,5 +77,27 @@ export const removeLinkFromCollection = async (collectionName, id, user) => {
...
+    const linksQuery = query(linksRef, where("__name__", "in", ids));
+    const linksDocs = await getDocs(linksQuery);
+    const parsedDocs = linksDocs.docs.map((document) => {
+        return {
+            title: document.data().title, 
+            url: document.data().url,
+        };
+    });
+    return parsedDocs || [];
+}
```
Comment (johnylab) [inline src/firebase/userCollections.js:113]: No changes are needed for this file. See comments above.

I would move this logic to `src/hooks/useUserLikes.ts`

File: src/hooks/useUserLikes.ts [hunk 12]
```diff
@@ -0,0 +1,28 @@
...
+        if (user) {
+            setInitialLikes(null);
+            getUserLikes();
+        };
+    }, [user]);
+
+    return { initialLikes };
+}
+
+export default useUserLikes;
```
Comment (johnylab) [inline src/hooks/useUserLikes.ts:37]: As commented above, please use Firebase methods to update `likes` automaticaly here and revert changes related to GlobalContext.

**Review by johnylab — CHANGES_REQUESTED** [review]

File: src/hooks/useUserLikes.ts (hunk 12 above)
Comment (kevynfg) [reply src/hooks/useUserLikes.ts:37]: Do you mean, use oneSnapShot from firestore with GlobalContext just like the "Stats" method ?

File: src/components/FavoritesSidebar/styled.js (hunk 2 above)
Comment (kevynfg) [reply src/components/FavoritesSidebar/styled.js]: Alright, changed to h2

Comment (johnylab) [reply src/components/FavoritesSidebar/styled.js]: have you checked the page headings structure?

File: src/hooks/useUserLikes.ts (hunk 12 above)
Comment (johnylab) [reply src/hooks/useUserLikes.ts:37]: I mean we don't need GlobalContext at all for this update. Please take a look at the comments above (possibly hidden):
image

Comment (johnylab) [reply src/hooks/useUserLikes.ts:37]: Especially this comment: https://github.com/agenciagift/feed/pull/7/files#r1178584482

File: src/App.js [hunk 13]
```diff
@@ -1,17 +1,22 @@
...
 import AddLink from './components/AddLink';
 import Header from './components/Header';
 import { Wrapper } from './components/layout';
 import LinkList from './components/LinkList';
+import FavoritesPanel from "./components/FavoritesLink";
+import { GlobalContext } from "./context/GlobalContext";
 
 function App() {
+  const { isSidebarOpen } = useContext(GlobalContext);
+  console.log('isSidebarOpen', isSidebarOpen);
```
Comment (johnylab) [inline src/App.js]: remove console

File: src/components/FavoritesLink/index.js [hunk 14]
```diff
@@ -1,25 +1,23 @@
...
-    const [showSidebar, setShowSidebar] = useState(false);
+export default function FavoritesPanel() {
     const { user } = useAuth();
     const { initialLikes } = useUserLikes(user);
-
+    const { isSidebarOpen } = useContext(GlobalContext);
     return !user || !initialLikes ? null : (
         <>
-            <FavoritesLink onClick={() => setShowSidebar(!showSidebar)}>
+            <FavoritesList>
```
Comment (johnylab) [inline src/components/FavoritesLink/index.js]: is it a list or a title? Looks like the result will be:

```
<ul>Favoritos</ul>
```

File: src/components/FavoritesSidebar/styled.js [hunk 15]
```diff
@@ -36,20 +36,20 @@ export const FavoritesUl = styled.ul`
...
 `
 
-export const FavoritesTitle = styled.h1`
+export const SidebarTitle = styled.h2`
     text-align: center;
     color: ${grey.light};
 `
 
-export const FavoritesLink = styled.a`
+export const SidebarLinks = styled.a`
```
Comment (johnylab) [inline src/components/FavoritesSidebar/styled.js]: this component is a single link, so the name should be singular: `SidebarLink`

File: src/components/UserMenu/index.js [hunk 16]
```diff
@@ -33,14 +36,26 @@ const UserMenu = ({ showMenu, setShowMenu }) => {
...
+    }
+
     return (
         <UserMenuContainer>
             {user
                 ? (
                     <>
                         <UserMenuText>
-                            {user.displayName}
+                            {firstNameSecondNameCapital(user.displayName)}
```
Comment (johnylab) [inline src/components/UserMenu/index.js]: 1. Have in mind you're editing the way the user defines their name in the login provider. We shouldn't care how the user saved their own info. If you see an issue with your own name in tests, you should update your Google/Github profile.
2. This is out of the scope, so should be broken out to a separate PR and a separate ticket.

File: src/components/UserMenu/index.js [hunk 17]
```diff
@@ -33,14 +36,26 @@ const UserMenu = ({ showMenu, setShowMenu }) => {
...
                         </UserMenuText>
+                        <UserMenuButton onClick={() => toggleSidebar()}>
```
Comment (johnylab) [inline src/components/UserMenu/index.js]:
```suggestion
                        <UserMenuButton onClick={toggleSidebar}>
```

File: src/components/FavoritesLink/styled.js [hunk 18]
```diff
@@ -0,0 +1,31 @@
...
+
+  to {
+    opacity: 1;
+  }
+`;
+
+export const FavoritesSidebarContainer = styled.div`
+    opacity: ${props => props.isSidebarOpen ? '1' : '0'};
+    transition: opacity 0.5s ease-in-out;
+    animation: ${fadeIn} 0.5s ease-in-out;
```
Comment (johnylab) [inline src/components/FavoritesLink/styled.js]: animation and transition together is redundant. You could pick one or the other instead of both

File: src/components/FavoritesSidebar/index.js [hunk 19]
```diff
@@ -0,0 +1,27 @@
...
+import { GlobalContext } from "../../context/GlobalContext";
+
+export default function SidebarPanel() { 
+    const { likes, toggleSidebar } = useContext(GlobalContext);
+
+    return likes && likes.length > 0 && (
+        <Sidebar>
+            <SidebarTitle>Favoritos</SidebarTitle>
+            <CloseSidebarButton onClick={() => toggleSidebar()}/>
+                <SidebarContent>
```
Comment (johnylab) [inline src/components/FavoritesSidebar/index.js:14]: please format the code (Alt + Shift + F in vscode for Windows)

File: src/components/Header/styled.js [hunk 20]
```diff
@@ -13,4 +13,5 @@ export const HeaderWrapper = styled(Wrapper)`
     display: flex;
     justify-content: space-between;
     align-items: center;
+    gap: ${MEDIUM};
 `;
```
Comment (johnylab) [inline src/components/Header/styled.js]: are these changes still necessary?

File: src/components/FavoritesLink/index.js (hunk 14 above)
Comment (kevynfg) [reply src/components/FavoritesLink/index.js]: my bad, i forgot to remove from the code

File: src/components/FavoritesLink/styled.js (hunk 18 above)
Comment (kevynfg) [reply src/components/FavoritesLink/styled.js]: Alright, i removed transition and kept animation

File: src/components/FavoritesSidebar/styled.js (hunk 15 above)
Comment (kevynfg) [reply src/components/FavoritesSidebar/styled.js]: done

File: src/components/Header/styled.js (hunk 20 above)
Comment (kevynfg) [reply src/components/Header/styled.js]: since we moved favorites to user avatar menu, i'll remove it thanks for the tip

File: src/components/UserMenu/index.js (hunk 16 above)
Comment (kevynfg) [reply src/components/UserMenu/index.js]: that's correct, i didnt think about it until you said it, it makes alot of sense. i'll remove this part of the code

File: src/hooks/useUserLikes.ts (hunk 12 above)
Comment (kevynfg) [reply src/hooks/useUserLikes.ts:37]: i can't find these comments if you could post here a print or something please.

File: src/firebase/userCollections.js (hunk 11 above)
Comment (kevynfg) [reply src/firebase/userCollections.js:113]: Made those changes, could you take a look ?

File: src/components/FavoritesSidebar/styled.js (hunk 2 above)
Comment (kevynfg) [reply src/components/FavoritesSidebar/styled.js]: Yes i did, do you mean to reuse the HeaderContainer already created with the SidebarTitle styling ?

**Edit**: something like
```
HeaderContainer
  <SidebarTitle> Favoritos </SidebarTitle>
HeaderContainer
```

Comment (johnylab) [reply src/components/FavoritesSidebar/styled.js]: I mean having the proper heading level according to the page structure.

File: src/hooks/useUserLikes.ts (hunk 12 above)
Comment (johnylab) [reply src/hooks/useUserLikes.ts:37]: Let me know if you can't see this:

https://user-images.githubusercontent.com/1944417/235928410-10ed2452-9b02-4459-829b-6307dd51419d.mov

Comment (kevynfg) [reply src/hooks/useUserLikes.ts:37]: Thanks for the vid really helped!

**Comment by kevynfg** [comment]
Made a few corrections to the PR, i couldn't fix the opacity animation to work like i expected to, it's just popping in the screen without keyframe animation. If anyone knows how to fix it i'll gladly appreciate

File: src/components/FavoritesSidebar/styled.js [hunk 21]
```diff
@@ -36,9 +78,13 @@ export const SidebarContent = styled.ul`
     flex-direction: column;
 `
 
-export const SidebarTitle = styled.h2`
+export const SidebarTitle = styled.header`
```
Comment (johnylab) [inline src/components/FavoritesSidebar/styled.js]: could possibly be `h3` or even `p`

File: src/components/FavoritesLink/index.js [hunk 22]
```diff
@@ -2,21 +2,48 @@ import { FavoritesSidebarContainer } from "./styled";
...
+                const userLinks = snap.data().items;
+                getAllUserLinksFromCollection(userLinks).then((links) => {
+                    if (links && links.length) setLikes(links)
+                });
+            });
+        }
+        return () => {
+            if (unsubscribe) unsubscribe();
+        }
+    }, [reference])
```
Comment (johnylab) [inline src/components/FavoritesLink/index.js]:
```suggestion
    useEffect(() => {
        if (!reference) {
            return;
        }
        
            return onSnapshot(reference, (snap) => {
                if (!snap.data().items) {
                    return;
                }
                const userLinks = snap.data().items;
                getAllUserLinksFromCollection(userLinks).then((links) => {
                    if (links && links.length) setLikes(links)
                });
            });
    }, [reference])
```

1. return faster and avoid hadouken code https://robsonalves.dev/7-simple-tips-to-write-a-better-code/
2. onSnapshot should (if I don't miss anything) always return the unsubs function or undefined

**Review by johnylab — CHANGES_REQUESTED** [review]

File: src/components/FavoritesSidebar/styled.js (hunk 2 above)
Comment (kevynfg) [reply src/components/FavoritesSidebar/styled.js]: I already changed, can you take a look? if it still wrong i could accept any suggestion

File: src/components/FavoritesSidebar/styled.js (hunk 21 above)
Comment (kevynfg) [reply src/components/FavoritesSidebar/styled.js]: changed to h3 then, thanks for the tip

**Review by johnylab — APPROVED** [review]
